pkg/manager/suspender/suspender.go

```diff
@@ -0,0 +1,244 @@
+		//
+		// NOTE: For tidb cluster and dm cluster, operator delete sts for all components one by one,
+		// so if any pod or sts deletion fails, the process will be stuck.
+		klog.Infof("suspend statefulset %s/%s for component %s", ns, stsName, ctx.ComponentID())
+
+		foreground := metav1.DeletePropagationForeground
+		err = s.deps.KubeClientset.AppsV1().StatefulSets(ns).Delete(context.TODO(), stsName,
+			metav1.DeleteOptions{PropagationPolicy: &foreground})
+		if err != nil {
+			return fmt.Errorf("failed to delete sts %s/%s: %s", ns, stsName, err)
+		}
+	} else {
+		// clear the status after the sts is deleted actually, so that we can
+		// use the `status.statefulset` is nil to determine whether sts suspension is done.
+		ctx.status.SetSynced(false)
+		ctx.status.SetStatefulSet(nil)
+		switch ctx.component {
+		case v1alpha1.PDMemberType:
+			ctx.status.(*v1alpha1.PDStatus).Members = nil
+			ctx.status.(*v1alpha1.PDStatus).Leader = v1alpha1.PDMember{}
+		case v1alpha1.TiDBMemberType:
+			ctx.status.(*v1alpha1.TiDBStatus).Members = nil
+		case v1alpha1.TiKVMemberType:
+			ctx.status.(*v1alpha1.TiKVStatus).Stores = nil
+		case v1alpha1.TiFlashMemberType:
+			ctx.status.(*v1alpha1.TiFlashStatus).Stores = nil
+		case v1alpha1.PumpMemberType:
+			ctx.status.(*v1alpha1.PumpStatus).Members = nil
+		case v1alpha1.TiCDCMemberType:
+			ctx.status.(*v1alpha1.TiCDCStatus).Captures = nil
+		case v1alpha1.DMMasterMemberType:
+			ctx.status.(*v1alpha1.MasterStatus).Members = nil
+			ctx.status.(*v1alpha1.MasterStatus).Leader = v1alpha1.MasterMember{}
+		case v1alpha1.DMWorkerMemberType:
+			ctx.status.(*v1alpha1.WorkerStatus).Members = nil
+		}
+	}
+
+	return nil
+}
+
+func (s *suspender) begion(ctx *suspendComponentCtx) error {
+	status := ctx.status
+	phase := v1alpha1.SuspendPhase
+	klog.Infof("begin to suspend component %s and transfer phase from %s to %s",
+		ctx.ComponentID(), status.GetPhase(), phase)
+	ctx.status.SetPhase(phase)
+	return nil
+}
+
+func (s *suspender) end(ctx *suspendComponentCtx) error {
+	status := ctx.status
+	phase := v1alpha1.NormalPhase
+	klog.Infof("end to suspend component %s and transfer phase from %s to %s",
+		ctx.ComponentID(), status.GetPhase(), phase)
+	ctx.status.SetPhase(phase)
+	return nil
+}
+
+// needsSuspendComponent returns whether suspender needs to to suspend the component
+func needsSuspendComponent(cluster v1alpha1.Cluster, comp v1alpha1.MemberType) bool {
+	spec := cluster.ComponentSpec(comp)
+	if spec == nil {
+		return false
+	}
+	action := spec.SuspendAction()
+	if action == nil {
+		return false
+	}
+
+	if action.SuspendStatefulSet {
+		return true
+	}
+
+	return false
+}
+
+// canSuspendComponent checks whether suspender can start to suspend the component
+func canSuspendComponent(cluster v1alpha1.Cluster, comp v1alpha1.MemberType) (bool, string) {
+	// only support to suspend Normal or Suspend cluster
+	if !cluster.ComponentIsNormal(comp) && !cluster.ComponentIsSuspending(comp) {
```

Comment:
From the view of TiDB Operator, we should be able to suspend cluster at any phase, right?